org-node/src/webserver.rs

```diff
@@ -0,0 +1,170 @@
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+use futures_util::{SinkExt, StreamExt, TryFutureExt};
+use librad::{git::Urn, PeerId};
+use serde::{Deserialize, Serialize};
+use tokio::sync::RwLock;
+use tokio::{
+    sync::mpsc::{self},
+    task::spawn,
+};
+use tokio_stream::wrappers::UnboundedReceiverStream;
+use warp::{
+    ws::{Message, WebSocket, Ws},
+    Filter,
+};
+
+/// Connected websocket client sender handles mapped by SocketAddr.
+type ConnectedWebSocketClients = Arc<RwLock<HashMap<SocketAddr, mpsc::UnboundedSender<WsEvent>>>>;
+
+/// Message type for establishing websocket sender.
+type EstablishWebSocketSender = (SocketAddr, mpsc::UnboundedSender<WsEvent>);
+
+// +----------------+
+// |                |    Receiver<WsEvent>
+// |  Main Process  +---------------------------+
+// |                |                           |
+// +-------+--------+                           |
+//         |                                    |
+//         | Sender<WsEvent>                    |
+//         |                                    |
+// +-------v--------+     +---------------------+---------------+-----+
+// |                |     |  Web Socket Process                 |     |
+// |   Client       |     |  (Spawned)                          |     |
+// |   Process      |     |                                     |     |
+// |   (Spawned)    |     |  Sender<WsEvent>                    |     |
+// |                |     |  Sender<EstablishWebSocketSender>   |     |
+// +----------------+     |                                     |     |
+//                        |  +----------------------------------v--+  |
+//                        |  | Connected Clients (Spawned)         |  |
+//                        |  |                                     |  |
+//                        |  | HashMap<SocketAddr,Sender<WsEvent>> |  |
+//                        |  | Receiver<WsEvent>                   |  |
+//                        |  | Receiver<EstablishWebSocketSender>  |  |
+//                        |  +-----^-+------------+-------------+--+  |
+//                        |        | |            |             |     |
+//                        +---+----+-+------+-----+---------+---+-----+
+//                            |      |      |     |         |   |
+//                        +---+------v+  +--+-----v---+  +--+---v-----+
+//                        |           |  |            |  |            |
+//                        | WS Client |  | WS Client  |  | WS Client  |
+//                        +-----------+  +------------+  +------------+
+
+/// Websocket server endpoint, e.g. ws://0.0.0.0:8336/subscribe
+pub const WEBSOCKET_PATH: &str = "subscribe";
+
+/// WebSocket event enum type to broadcast to connected websocket peers.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum WsEvent {
+    UpdatedRef {
+        urn: Urn,
+        // Rev is Rev::Git(git2::Oid)
+        // Using string to allow for serialization
+        rev: String,
```

Comment:
Can we call this `oid` if it's an oid? Got kinda confused thinking it was ` Rev`

Reply:
commit 0452b3c fixes.